### ISSUE TITLE
Tunnel connection with closed port

### ISSUE THREAD

**Issue by littleskunk** [Storj/core#223]
### Package Versions

Replace the values below using the output from `npm list storj`.

```
Storj Share 0.8.6 
Core Library 1.3.3 
Protocol 0.7.0
```

Replace the values below using the output from `node --version`.

```
v4.4.5
```

### Expected Behavior

Please describe the program's expected behavior. Include an example of your
usage code in the back ticks below if applicable.

```
I should receive contract PUBLISH messages.
```

### Actual Behavior

Please describe the program's actual behavior. Please include any stack traces
or log output in the back ticks below.

GUI node running and open tunnel connection via bridge node.
```
{info} [Wed Jun 29 2016 11:16:18 GMT+0200 (Mitteleuropäische Sommerzeit)]
node created with nodeID 310771213cb3cb79c3dd1cdce6ef673b1196fbac
{info} [Wed Jun 29 2016 11:16:20 GMT+0200 (Mitteleuropäische Sommerzeit)]
tunnel successfully established: {"address":"104.196.135.20","port":41953}
```

Lets use this information to connect my cli node.
```
{"level":"info","message":"node created with nodeID b408aeb67291e4d1994a95394e674c09eb683d3c","timestamp":"2016-06-29T09:17:40.220Z"}
{"level":"info","message":"attempting to join network via storj://104.196.135.20:41953/310771213cb3cb79c3dd1cdce6ef673b1196fbac","timestamp":"2016-06-29T09:17:40.224Z"}
{"level":"debug","message":"entering overlay network via {\"protocol\":\"0.7.0\",\"address\":\"104.196.135.20\",\"port\":41953,\"nodeID\":\"310771213cb3cb79c3dd1cdce6ef673b1196fbac\",\"lastSeen\":1467191860233}","timestamp":"2016-06-29T09:17:40.234Z"}
{"level":"debug","message":"updating contact {\"protocol\":\"0.7.0\",\"address\":\"104.196.135.20\",\"port\":41953,\"nodeID\":\"310771213cb3cb79c3dd1cdce6ef673b1196fbac\",\"lastSeen\":1467191860233}","timestamp":"2016-06-29T09:17:40.242Z"}
{"level":"debug","message":"creating new bucket for contact at index 159","timestamp":"2016-06-29T09:17:40.243Z"}
{"level":"debug","message":"contact not in bucket, moving to head","timestamp":"2016-06-29T09:17:40.244Z"}
{"level":"debug","message":"searching for nodes close to key b408aeb67291e4d1994a95394e674c09eb683d3c","timestamp":"2016-06-29T09:17:40.249Z"}
{"level":"debug","message":"performing network walk for NODE b408aeb67291e4d1994a95394e674c09eb683d3c","timestamp":"2016-06-29T09:17:40.263Z"}
{"level":"debug","message":"starting contact iteration for key b408aeb67291e4d1994a95394e674c09eb683d3c","timestamp":"2016-06-29T09:17:40.264Z"}
{"level":"debug","message":"querying 310771213cb3cb79c3dd1cdce6ef673b1196fbac for key b408aeb67291e4d1994a95394e674c09eb683d3c","timestamp":"2016-06-29T09:17:40.268Z"}
{"level":"info","message":"sending FIND_NODE message to {\"protocol\":\"0.7.0\",\"address\":\"104.196.135.20\",\"port\":41953,\"nodeID\":\"310771213cb3cb79c3dd1cdce6ef673b1196fbac\",\"lastSeen\":1467191860270}","timestamp":"2016-06-29T09:17:40.270Z"}
{"level":"debug","message":"queuing callback for reponse to 9d09977d2894f91d99a1697335d0da3f1a4d0859","timestamp":"2016-06-29T09:17:40.374Z"}
{"level":"debug","message":"checking pending rpc callbacks for expirations","timestamp":"2016-06-29T09:18:10.141Z"}
{"level":"debug","message":"checking pending rpc callbacks for expirations","timestamp":"2016-06-29T09:18:40.170Z"}
{"level":"warn","message":"rpc call 9d09977d2894f91d99a1697335d0da3f1a4d0859 timed out","timestamp":"2016-06-29T09:18:40.171Z"}
{"level":"debug","message":"removing contact {\"protocol\":\"0.7.0\",\"address\":\"104.196.135.20\",\"port\":41953,\"nodeID\":\"310771213cb3cb79c3dd1cdce6ef673b1196fbac\",\"lastSeen\":1467191860270}","timestamp":"2016-06-29T09:18:40.171Z"}
{"level":"warn","message":"query failed, removing contact for shortlist, reason RPC with ID `9d09977d2894f91d99a1697335d0da3f1a4d0859` timed out","timestamp":"2016-06-29T09:18:40.171Z"}
{"level":"debug","message":"removing contact {\"protocol\":\"0.7.0\",\"address\":\"104.196.135.20\",\"port\":41953,\"nodeID\":\"310771213cb3cb79c3dd1cdce6ef673b1196fbac\",\"lastSeen\":1467191860265}","timestamp":"2016-06-29T09:18:40.172Z"}
{"level":"debug","message":"finished iteration, handling results","timestamp":"2016-06-29T09:18:40.173Z"}
{"level":"warn","message":"failed to connect to seed storj://104.196.135.20:41953/310771213cb3cb79c3dd1cdce6ef673b1196fbac","timestamp":"2016-06-29T09:18:40.173Z"}
{"level":"info","message":"publishing message on topic \"0e01\"","timestamp":"2016-06-29T09:18:40.177Z"}
{"level":"info","message":"subscribing to topic \"0e01\"","timestamp":"2016-06-29T09:18:40.178Z"}
{"level":"debug","message":"requesting neighbors' bloom filters","timestamp":"2016-06-29T09:18:40.180Z"}
{"level":"info","message":"subscribing to topic \"0e00\"","timestamp":"2016-06-29T09:18:40.181Z"}
{"level":"debug","message":"requesting neighbors' bloom filters","timestamp":"2016-06-29T09:18:40.182Z"}
 [error]  Failed to join the network
```

Portcheck (http://www.yougetsignal.com/tools/open-ports/)
Port 41953 is closed on 104.196.135.20

My GUI node is still waiting for incomming messages and for some reason it doesn't detect the closed port.

I try this a few times. The bridge gives me a tunnel port >30000 and they are all closed. I assume the port is out of range. Usualy the bridge is giving me a port <10000.

### Steps to Reproduce

Please include the steps the reproduce the issue, numbered below. Include as
much detail as possible.

1. Start node that needs a tunnel connection
2. Check the port

**Comment by bookchin** [comment]
@phutchins any ideas here? I don't think it's a core issue - no changes made to tunneling.

**Comment by phutchins** [comment]
@bookchin I see a large number of ports open well above the specified range in our configs. Our configs have 8500-8532 configured for the tunnel range. The arbitrary ports are 30XXX 40XXX 50XXX. I see no reason in the configs/infra for ports being assigned outside of the specified range.

**Comment by phutchins** [comment]
@bookchin & @littleskunk I've opened up ports in the arbitrary range to alleviate the symptom of the issue until we can track it down.

**Comment by bookchin** [comment]
Interesting, I bet it has to do with the renter minions getting their configs passed to them and the minions are not getting the gateway config properties and are defaulting to random ports.

**Comment by bookchin** [comment]
This issue is specific to bridge, see https://github.com/Storj/bridge/issues/128

**Comment by littleskunk** [comment]
I like the fast bridge fix and we realy need that. Please go on.

At the meantime I my next step is to setup my cli node and allow 100 tunnel connection. I will open farmer and tunnel server port but not the gateway ports.

**Comment by phutchins** [comment]
I'm curious to see if a single minion behaves the same way. We've restarted which clears the tunnels and I do see some on the configured port range.

**Comment by littleskunk** [comment]
I was able to reproduce this issue. -> Reopen

Tunnel server node has a bad network configuration. The port for the kad node and the tunnel server are open but the tunnel gateway ports are closed.

Tunnel client node is using that closed port for a tunnel connection. It will never receive any messages.
```
{"level":"info","message":"node created with nodeID 918d31888ba771013037f84efae94843683ff445","timestamp":"2016-06-30T15:59:31.335Z"}
{"level":"info","message":"tunnel successfully established: {\"address\":\"littleskunk.myqnapcloud.com\",\"port\":4003}","timestamp":"2016-06-30T16:00:08.080Z"}
```

Node 3 is using the tunnel client node as seed but as expected he can't send messages to a close port.
```
{"level":"info","message":"sending FIND_TUNNEL message to {\"protocol\":\"0.7.0\",\"address\":\"littleskunk.myqnapcloud.com\",\"port\":4003,\"nodeID\":\"918d31888ba771013037f84efae94843683ff445\",\"lastSeen\":1467302815596}","timestamp":"2016-06-30T16:06:55.596Z"}
{"level":"debug","message":"queuing callback for reponse to 00acee7b24bfbca8446a377ce8bdb46cbe57d9b2","timestamp":"2016-06-30T16:06:55.626Z"}
{"level":"warn","message":"missing or empty reply from contact","timestamp":"2016-06-30T16:06:56.637Z"}
{"level":"debug","message":"checking pending rpc callbacks for expirations","timestamp":"2016-06-30T16:07:25.706Z"}
{"level":"warn","message":"rpc call 00acee7b24bfbca8446a377ce8bdb46cbe57d9b2 timed out","timestamp":"2016-06-30T16:07:25.710Z"}
{"level":"debug","message":"removing contact {\"protocol\":\"0.7.0\",\"address\":\"littleskunk.myqnapcloud.com\",\"port\":4003,\"nodeID\":\"918d31888ba771013037f84efae94843683ff445\",\"lastSeen\":1467302815596}","timestamp":"2016-06-30T16:07:25.722Z"}
```

**Comment by bookchin** [comment]
@littleskunk try adding some logging around here https://github.com/Storj/core/blob/master/lib/network/protocol.js#L392

Find out if that `createPortMapping` method is getting called or not.

**Comment by bookchin** [comment]
Also make sure this is getting set https://github.com/Storj/core/blob/master/lib/network/transport.js#L65

If not, we can determine why

**Comment by littleskunk** [comment]
My router doesn't support UPnP. I have to open the ports myself and I am running my node with `forward=false` / `noforward=true`.

I am leaving here: https://github.com/Storj/core/blob/master/lib/network/transport.js#L62

**Comment by bookchin** [comment]
Ah, are you forwarding the gateway port range manually too? I suppose we may need to ask for another PROBE after establishing a tunnel to ensure it is working.

**Comment by littleskunk** [comment]
Yes they are open again. I closed them only for this test. A PROBE message will solve this situation :)

**Comment by bookchin** [comment]
Okay patch will come later today. Thanks.

**Comment by bookchin** [comment]
@littleskunk okay, want to try my fork? https://github.com/Storj/core/pull/234

**Comment by bookchin** [comment]
Ooh also probably need to do another network walk once a different tunnel is established otherwise node won't learn your new contact information. 

I bet that's what is happening in #221 too.

**Comment by littleskunk** [comment]
```
C:\Users\jensh_000\AppData\Roaming\npm
└─┬ storjshare-cli@4.0.4
  └── storj@1.3.6  (git://github.com/bookchin/storj-core.git#6e0ad5d22c0c1565e5944f542928ed501cae1807)
```

Tunnel client node did the same mistake. Connected with a closed port.
```
{"level":"info","message":"tunnel successfully established: {\"address\":\"littleskunk.myqnapcloud.com\",\"port\":4002}","timestamp":"2016-07-01T09:15:39.513Z"}
{"level":"info","message":"testing newly established tunnel { address: 'littleskunk.myqnapcloud.com', port: 4002 }","timestamp":"2016-07-01T09:15:39.538Z"}
{"level":"info","message":"sending PROBE message to {\"protocol\":\"0.7.0\",\"address\":\"192.168.123.101\",\"port\":4000,\"nodeID\":\"b408aeb67291e4d1994a95394e674c09eb683d3c\",\"lastSeen\":1467364539554}","timestamp":"2016-07-01T09:15:39.554Z"}
```

**Comment by littleskunk** [comment]
Sending the PROBE message to the tunnel server node will not work. My router has no NAT loopback. For my tunnel server node littleskunk.myqnapcloud.com is localhost. PROBE message response to the localhost gateway port will work even if the port is closed.

Is it possible for the tunnel client node to send a PROBE message to the gateway port?

**Comment by bookchin** [comment]
Unfortunately, no, it will behave the same way (as currently implemented). You will send a PROBE through your tunnel, the tunnel server will send it through the gateway (which is local) and it will still succeed.

We could perform a PROBE ourselves though instead of requesting one. So the tunnel client will simply try to talk to itself through the newly established gateway.

**Comment by bookchin** [comment]
Okay, @littleskunk - I added a `ContactChecker` class that will get used instead of PROBE request after establishing a tunnel.

It just attempts to open a socket to itself using the gateway. If the connection times out or is refused, then we establish a different tunnel.

**Comment by littleskunk** [comment]
```
C:\Users\jensh_000\AppData\Roaming\npm
└─┬ storjshare-cli@4.0.4
  └── storj@1.3.6  (git://github.com/bookchin/storj-core.git#223de7812c63ecc228bfd0e28cf3588d4e2c18af)
```

Finding a new tunnel is not working.
```
{"level":"info","message":"tunnel successfully established: {\"address\":\"littleskunk.myqnapcloud.com\",\"port\":4002}","timestamp":"2016-07-01T17:40:49.851Z"}
{"level":"info","message":"testing newly established tunnel { address: 'littleskunk.myqnapcloud.com', port: 4002 }","timestamp":"2016-07-01T17:40:49.904Z"}
{"level":"warn","message":"tunnel test failed, establishing new tunnel","timestamp":"2016-07-01T17:40:50.947Z"}
{"level":"warn","message":"tunnel connection closed","timestamp":"2016-07-01T17:40:50.947Z"}
{"level":"info","message":"sending OPEN_TUNNEL message to {\"protocol\":\"0.7.0\",\"address\":\"192.168.2.110\",\"port\":23130,\"nodeID\":\"71c099a51fa533938481bf1ccb4b4c73b11848da\",\"lastSeen\":1467394850962}","timestamp":"2016-07-01T17:40:50.962Z"}
{"level":"debug","message":"queuing callback for reponse to 9bacdd13e4ac6fd63c319d72d207aecf95a18009","timestamp":"2016-07-01T17:40:51.050Z"}
{"level":"warn","message":"tunnel connection lost, reason: Tunnel is not open","timestamp":"2016-07-01T17:40:51.066Z"}
 [error]  Failed to establish tunnel, reason: No tunnelers were returned
```

**Comment by littleskunk** [comment]
```
C:\Users\jensh_000\AppData\Roaming\npm
└─┬ storjshare-cli@4.0.4
  └── storj@1.3.6  (git://github.com/bookchin/storj-core.git#223de7812c63ecc228bfd0e28cf3588d4e2c18af)
```

My node is no connected and receiving messages. Perfect :)
